internal/extensions/extensions.go

```diff
@@ -0,0 +1,238 @@
+		return fmt.Errorf("Internal extension name %q is invalid: Extension name must contain only lowercase letters and underscores, and must begin by `internal:`", extension)
+	}
+
+	return nil
+}
+
+// HasExtension reports whether the extension set supports the given extension.
+func (e *Extensions) HasExtension(ext string) bool {
+	for _, extension := range *e {
+		if extension == ext {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Version returns the number of extensions in the set, representing its version number.
+func (e *Extensions) Version() int {
+	return len(*e)
+}
+
+// NewExtensionRegistry creates a new instance of the Extensions struct, which represents a registry of extensions.
+// We have the option to populate the internal extensions or not. Having s option becomes handy
+// when we want to create a custom registry of extensions (see `NewExtensionRegistryFromList`)
+func NewExtensionRegistry(populateInternal bool) (*Extensions, error) {
+	extensions := new(Extensions)
+	var err error
+	if populateInternal {
+		for _, internalExtension := range *internalExtensions {
+			err = validateInternalExtension(internalExtension)
+			if err != nil {
+				return nil, err
+			}
+
+			i := sort.Search(len(*extensions), func(i int) bool {
+				return (*extensions)[i] >= internalExtension
+			})
+
+			*extensions = append(*extensions, internalExtension)
+			copy((*extensions)[i+1:], (*extensions)[i:])
+			(*extensions)[i] = internalExtension
+		}
+	}
+
+	return extensions, nil
+}
+
+// NewExtensionRegistryFromList creates a new instance of the Extensions struct, from a list of extensions.
+// The passed extensions could be internal or external or just invalid.
+// The function is typically used to create a registry from an API response in order to compare
+// the source and the target systems' extensions for compatibility.
+// The returned registry is always sorted with internal extensions first.
+func NewExtensionRegistryFromList(extensions []string) (*Extensions, error) {
+	registry, err := NewExtensionRegistry(false)
+	if err != nil {
+		return nil, err
+	}
+
+	var isInternal bool
+	for _, extension := range extensions {
+		err1 := validateExternalExtension(extension)
+		if err1 != nil {
+			// The extension could be an internal extension.
+			err2 := validateInternalExtension(extension)
+			if err2 != nil {
+				// The extension is invalid in both case.
+				return nil, fmt.Errorf("Extension %q is invalid", extension)
+			}
+
+			isInternal = true
+		} else {
+			isInternal = false
+		}
+
+		i := sort.Search(len(*registry), func(i int) bool {
+			currentIsInternal := validateInternalExtension((*registry)[i]) == nil
+			if isInternal {
+				return !currentIsInternal || (*registry)[i] >= extension
+			}
+
+			return !currentIsInternal && (*registry)[i] >= extension
+		})
+
+		*registry = append(*registry, extension)
+		copy((*registry)[i+1:], (*registry)[i:])
+		(*registry)[i] = extension
+	}
+
+	return registry, nil
```

Comment:
Same here, it's not clear to me why we need to have all of this, and can't just directly append to `registry` and return.

Reply:
same reason as above

Reply:
So here, it looks like the lists are also sorted alphanumerically, but we still have a preference for `internal:` in the front, which means the full sort method we have is "alphanumeric, except for the prefix `internal:` which is front-loaded"

This makes it even more confusing to me why "alphanumeric" is helpful here.

Reply:
fixed